### DISCUSSION TITLE
DuplicateIDs error on latest build when swapping widgets

### DISCUSSION THREAD

**Discussion by AL-Troy** [Textualize/textual#5227] [Q&A]
I've recently upgraded one of my projects to the latest build and I'm running into issues switching between widgets. I have a central_widget that serves as a ScrollableContainer, and I switch out a Static with ID #main_widget as the application moves through different stages in our process. The following implementation worked normally on <0.40.0, but now throws DuplicateID errors on 0.85.2:

`async def unload_and_load_widget(self, widget: Static):
        """Using the application framework load a widget

        Args:
            widget (Any): The widget to load
        """
        self.log.debug(f"Loading: {widget}")
        central_widget = self.view.query_one("#central_widget")
        try:
            widgets = central_widget.query("#main_widget")
            current_widget = widgets.last()
        except NoMatches:
            current_widget = None

        if current_widget is not None:
            self.log.debug(f"Removing: {current_widget}")
            await current_widget.remove()

        await central_widget.mount(widget)`

I've tried awaiting sleep() to wait for the remove to finish, remove_children() instead of remove(), or waiting until central_widget.message_queue_size hits 0, but nothing seems to work.

**Comment by willmcgugan** [comment]
Hard to say without an MRE or traceback, but `DuplicateIds` is raised when you attempt to mount a widget and that ID already exists in children. If you print `central_widget.children` you might be able to see the duplicate.

Are you perhaps re-using a widget, after it has been removed from the DOM? That is unlikely to work.

There are likely better ways of implementing swapping widgets, if you have leaped from .040.0 to 0.85.0. Perhaps by adding all the widgets and setting there `display` attribute as required, or with a [ContentSwitcher](https://textual.textualize.io/widget_gallery/#contentswitcher)

**Reply by AL-Troy** [reply]
I'll try to put together an MRE that could help clarify the issue. 

When the application starts, `MyWidget(id="main_widget")` is immediately added as a child to the empty central_widget, and then sometime later we `await MyWidget.remove()` so that we can mount `SomeOtherWidget(id="main_widget")`. I can see during debugging that central_widget.children is still populated with the original `MyWidget` even after `remove()` is awaited and all messages in central_widget queue have been processed. Since the original widget is still in place, the duplicate IDs error makes sense, I'm just not sure why it's still around after awaiting `remove()`.

I could try having all the widgets added and modifying `display` or using ContentSwitcher. Do you think either of these solutions have much of a performance impact while running on a lower-end machine like an RPi 3?
